src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -2248,6 +2286,148 @@ public static (A, A) GetTuple()
+        => throw null;
+}
+namespace System
+{
+    public struct ValueTuple<T1, T2>
+        where T1 : class
+        where T2 : class
+    {
+        public T1 Item1;
+        public T2 Item2;
+
+        public ValueTuple(T1 item1, T2 item2)
+        {
+            this.Item1 = item1;
+            this.Item2 = item2;
+            throw null;
+        }
+    }
+}
+";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (6,30): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T1' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (0, 1); // constraint violated by tuple in source
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "0").WithArguments("System.ValueTuple<T1, T2>", "T1", "int").WithLocation(6, 30),
+                // (6,33): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T2' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (0, 1); // constraint violated by tuple in source
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "1").WithArguments("System.ValueTuple<T1, T2>", "T2", "int").WithLocation(6, 33),
+                // (6,30): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T1' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (0, 1); // constraint violated by tuple in source
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "0").WithArguments("System.ValueTuple<T1, T2>", "T1", "int").WithLocation(6, 30),
+                // (6,33): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T2' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (0, 1); // constraint violated by tuple in source
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "1").WithArguments("System.ValueTuple<T1, T2>", "T2", "int").WithLocation(6, 33),
+                // (7,30): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T1' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (this, this); // constraint violated by converted tuple
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "this").WithArguments("System.ValueTuple<T1, T2>", "T1", "int").WithLocation(7, 30),
+                // (7,36): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T2' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (this, this); // constraint violated by converted tuple
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "this").WithArguments("System.ValueTuple<T1, T2>", "T2", "int").WithLocation(7, 36)
```

Comment:
Why is `this` converted to `int` rather than comparing using reference equality?

Reply:
From our discussion that's existing binary operator behavior. `this == this` behaves the same (and involves the `==(C, int)` operator, rather than reference equality).

---
In reply to: [173873152](https://github.com/dotnet/roslyn/pull/25374#discussion_r173873152) [](ancestors = 173873152)